### PR TITLE
add spec.networkPluginMode to AzureManagedControlPlane

### LINKED ISSUE #3678: Add support for Azure CNI Overlay on AKS managed clusters
/kind feature

**Describe the solution you'd like**
Add support for [Azure CNI Overlay](https://learn.microsoft.com/en-us/azure/aks/azure-cni-overlay) on AKS managed clusters.

**Anything else you would like to add:**
Microsoft.ContainerService/managedClusters API implements this with:

[`networkProfile.networkPluginMode: overlay`](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/get?tabs=HTTP#networkpluginmode)
[`networkProfile.podCidr: <cidr>`](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/get?tabs=HTTP#containerservicenetworkprofile)

### PR DESCRIPTION
**Pull request by nojnhuh**
**What type of PR is this?**
/kind feature
/area managedclusters

**What this PR does / why we need it**: This PR exposes the `networkProfile.networkPluginMode` AKS API field from CAPZ at AzureManagedControlPlane's `spec.networkPluginMode`.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3678

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [X] squashed commits
- [ ] includes documentation
- [X] adds unit tests

**Release note**:

```release-note
Added `spec.networkPluginMode` to AzureManagedControlPlane
```

### CODE REVIEW

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (Jont828) [inline api/v1beta1/azuremanagedcontrolplane_types.go:126]: I assume this means that there's only one possible value? In that case, what happens with the `// +optional` and `omitempty` tags?

Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_types.go:126]: These allowed values are only enforced when the value is non-nil, so this is the behavior:
```
networkPluginMode: "overlay" # ok

networkPluginMode: null # ok

networkPluginMode: "" # error
```

File: azure/services/managedclusters/spec_test.go
Comment (jackfrancis) [inline azure/services/managedclusters/spec_test.go:384]: Is this a valid use case? (i.e., does AKS allow us to change this value to an existing cluster?)

Comment (nojnhuh) [reply azure/services/managedclusters/spec_test.go:384]: Yes: https://learn.microsoft.com/en-us/azure/aks/azure-cni-overlay#upgrade-an-existing-cluster-to-cni-overlay

Comment (jackfrancis) [reply azure/services/managedclusters/spec_test.go:384]: Are there any CAPZ apiserver webhooks we can add to protect these potential failures?

- Doesn't use the dynamic pod IP allocation feature.
- Doesn't have network policies enabled.
- Doesn't use any Windows node pools with docker as the container runtime.

(i.e., does CAPZ support the above features that don't work w/ Azure CNI overlay)

Comment (nojnhuh) [reply azure/services/managedclusters/spec_test.go:384]: Where would we check for each of those things? If they're defined in CAPZ objects then we could check those. I'm not sure if we have plumbing set up in the webhooks to do Azure API calls though.

Comment (jackfrancis) [reply azure/services/managedclusters/spec_test.go:384]: Right, we would only do this effectively if there was a CAPZ representation that corrrelated to the violating AKS features. If we haven't yet implemented support for the above then we would probably not want to mess around with pre-validating against the AKS API mid-flight (we should defer to AKS on that and improve the UX by actually implementing CAPZ wrapping of those features, IMO).

We do seem to support NetworkPolicy in CAPZ, so maybe we could add a webhook to enforce that we don't ship that config w/ CNI overlay.

Comment (nojnhuh) [reply azure/services/managedclusters/spec_test.go:384]: I can add a check for network policy. I don't think we can (easily) add anything for those other items though.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (nojnhuh) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:550]: I verified that AKS does allow this on create. It only errors on updates.

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (Jont828) [reply api/v1beta1/azuremanagedcontrolplane_types.go:126]: Gotcha, are there any other cases where we do validation like this instead of with the webhook? I'm fine to proceed as is, but figured it's worth noting that it might be confusing to have validation in more than one place.

**Comment by Jont828** [comment]
LGTM pending squash

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_types.go:126]: Not sure what you mean? The kubebuilder annotation here is the only thing defining this specific validation.

**Comment by nojnhuh** [comment]
Just pushed a change to improve CIDR defaults in the webhook when overlay is enabled so they match AKS's defaults. Thank you @felfa01 for the feedback in https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3678#issuecomment-1715680082!

File: api/v1beta1/azuremanagedcontrolplane_types.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_types.go:126]: Yeah, I think the general principle is

1. Prefer yaml schema validation (this is limited, but if your validation can be express in terms of kubebuilder pragmas then this is better)
2. Only do one or the other (both is worse)
3. When validation is trickier, use webhooks

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:550]: Wut

We're saying:

1. You can create a cluster w/ CNI Overlay + NetworkPolicy when you create your cluster
2. You cannot update a cluster w/ NetworkPolicy configured from non-CNI Overlay to CNI Overlay
3. You can update a cluster to CNI Overlay if NetworkPolicy is not enabled

?

Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:550]: Correct

**Comment by nojnhuh** [comment]
AKS test failing due to #3970

**Comment by DavidSpek** [comment]
Should this be part of the [v1.11](https://github.com/kubernetes-sigs/cluster-api-provider-azure/milestone/33) milestone?

**Comment by nojnhuh** [comment]
@DavidSpek Yes, the plan is to get this in before the release.

/milestone v1.11

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go:550]: I wonder if scenario #1 actually works? Probably out-of-scope for this PR to definitely prove that, so we can take the API at its word and implement CAPZ accordingly.

Thanks for doing the extra validation here.

**Comment by nojnhuh** [comment]
@jackfrancis @Jont828 Any other thoughts here? Looking to include this in the v1.11 release today if possible.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (mboersma) [inline api/v1beta1/azuremanagedcontrolplane_webhook.go:733]:
```suggestion
// validateNetworkPluginMode validates a NetworkPluginMode.
```

**Review by mboersma — COMMENTED** [review]
lgtm, all I could find was a code comment discrepency.

File: api/v1beta1/azuremanagedcontrolplane_webhook.go
Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go]: caught red-handed lol. Fixed!

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_webhook.go]: @mboersma you win this round

File: api/v1beta1/azuremanagedcontrolplane_default.go
Comment (jackfrancis) [inline api/v1beta1/azuremanagedcontrolplane_default.go:84]: One final validation possibility here... so I see we're assigning an overlay-specific CIDR value for VNET and subnet if the user doesn't declare one. Is there any validation we can add here for the overlay CIDRs? Does AKS require that we use particular nets here?

We're already doing some basic hygiene in `validateSubnetCIDR` and `validateVnetCIDR` (`api/v1beta1/azurecluster_validation.go`) so we get that for free.

Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_default.go:84]: Are you suggesting to call the same `validateVnetCIDR` and `validateSubnetCIDR` on these values?

> Does AKS require that we use particular nets here?

No AFAICT, these overlay-specific defaults match AKS's defaults when overlay is enabled per https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/3678#issuecomment-1715680082.

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_default.go:84]: I'm saying we're already calling those validate funcs, so that's good (as far as I can see we're reusing existing property CIDR fields w/ this implementation)

My question is what happens when a user *explicitly* passes in `"10.0.0.0/8"` (for example) as the VNET CIDR _and_ enables overlay CNI? Is AKS opinionated: "Hey, that CIDR block is reserved for non-overlay configurations!".

If we don't know all the answers to the above, I think it's fair to keep that out of scope, just curious.

Comment (nojnhuh) [reply api/v1beta1/azuremanagedcontrolplane_default.go:84]: > My question is what happens when a user _explicitly_ passes in `"10.0.0.0/8"` (for example) as the VNET CIDR _and_ enables overlay CNI? Is AKS opinionated: "Hey, that CIDR block is reserved for non-overlay configurations!".

The comment I linked above shows the error with the non-overlay defaults, and it wasn't that specific:
`NetworkProfile CIDR 10.244.0.0/16 overlaps with reserved VNET CIDR 10.0.0.0/8`

I don't think a user explicitly passing in the value would affect that since it results in the same config sent to AKS either way.

Comment (jackfrancis) [reply api/v1beta1/azuremanagedcontrolplane_default.go:84]: I think we can add more explicit CAPZ validations that mirror AKS validations over time if it has user value (in most cases failing the declarative spec is more user-friendly than looking for an API error in the CAPZ controller logs). Shouldn't block this change.

**Review by mboersma — APPROVED** [review]
/lgtm
/approve

**Comment by nojnhuh** [comment]
Still need to resolve impending conflicts with #3973 after that merges, so I'll need one more lgtm after that @mboersma @jackfrancis. Then I'll sqaush and push such that the lgtm sticks.

**Comment by nojnhuh** [comment]
@mboersma @CecileRobertMichon This should be ready for a final lgtm.

**Review by mboersma — COMMENTED** [review]
/lgtm

**Comment by nojnhuh** [comment]
squashed!
/hold cancel

**Pull request merged**
